File: packages/sdk-ts/src/oauth.ts
// SPDX-License-Identifier: Apache-2.0

// suite.oauth.* — OAuth-on-behalf-of-user operations.
//
// Lets backend agents act on behalf of the user in third-party APIs.
// The first shipped providers are Google and GitHub. The runtime owns
// the OAuth dance + vault; this SDK is the client surface.
//
// - authorizeUrl(provider, opts) -> string  : consent URL to redirect to
// - connected()                  -> ConnectionList
// - token(provider, opts)        -> string   : fresh access token (internal-only)
// - disconnect(provider, opts)   -> boolean : revoke + soft-delete
//
// `token()` sends an `X-AF-Stack-Internal: 1` header. The header is NOT
// in the runtime's CORS allow-list, so browsers cannot reach this
// endpoint — only server-side callers (workload modules, agents, this
// SDK from a Node process) can.

import { z } from "zod"
import { request, type HttpOptions } from "./_http.js"

export const ConnectionSchema = z.object({
  provider: z.string(),
  scopes: z.array(z.string()).default([]),
  connected_at: z.string(),
  expires_at: z.string().nullable().optional(),
})
export type Connection = z.infer<typeof ConnectionSchema>

export const ConnectionListSchema = z.object({
  connections: z.array(ConnectionSchema).default([]),
})
export type ConnectionList = z.infer<typeof ConnectionListSchema>

export const AccessTokenSchema = z.object({
  provider: z.string(),
  user_id: z.string(),
  access_token: z.string(),
  scopes: z.array(z.string()).default([]),
  expires_at: z.string().nullable().optional(),
})
export type AccessToken = z.infer<typeof AccessTokenSchema>

const AuthorizeResponseSchema = z.object({
  provider: z.string(),
  authorization_url: z.string(),
  redirect_uri: z.string(),
  scopes: z.array(z.string()).default([]),
})

export interface AuthorizeUrlOptions extends HttpOptions {
  scopes?: string[]
  returnTo?: string
}

export interface TokenOptions extends HttpOptions {
  userId?: string
}

export interface DisconnectOptions extends HttpOptions {
  userId?: string
}

/**
 * Return the provider's consent URL the user should be redirected to.
 *
 * The runtime signs an HMAC state nonce that round-trips through the
 * provider's consent page; on callback it's verified to defeat CSRF.
 * `returnTo` is validated against an allowed-origin list.
 */
export async function authorizeUrl(
  provider: string,
  opts: AuthorizeUrlOptions = {},
): Promise<string> {
  if (typeof provider !== "string" || provider.length === 0) {
    throw new Error("provider must be a non-empty string")
  }
  const { scopes, returnTo, ...http } = opts
  const body: Record<string, unknown> = {}
  if (scopes !== undefined) body.scopes = scopes
  if (returnTo !== undefined) body.return_to = returnTo
  const raw = await request<unknown>(
    "POST",
    `/oauth/${encodeURIComponent(provider)}/authorize`,
    body,
    http,
  )
  return AuthorizeResponseSchema.parse(raw).authorization_url
}

/** List active OAuth connections for the current user. */
export async function connected(opts: HttpOptions = {}): Promise<ConnectionList> {
  const raw = await request<unknown>("GET", "/oauth/connections", null, opts)
  return ConnectionListSchema.parse(raw)
}

/**
 * Return a fresh access token for `provider`. Auto-refreshes if the
 * stored token is within 60s of expiry.
 *
 * Internal-only: the runtime requires `X-AF-Stack-Internal: 1`. Browsers
 * cannot send this header cross-origin thanks to the CORS allow-list,
 * so only server-side callers reach it.
 */
export async function token(
  provider: string,
  opts: TokenOptions = {},
): Promise<string> {
  if (typeof provider !== "string" || provider.length === 0) {
    throw new Error("provider must be a non-empty string")
  }
  const { userId, headers, ...rest } = opts
  const merged: HttpOptions = {
    ...rest,
    headers: {
      ...(headers ?? {}),
      "X-AF-Stack-Internal": "1",
    },
  }
  const body: Record<string, unknown> = { provider }
  if (userId !== undefined) body.user_id = userId
  const raw = await request<unknown>("POST", "/oauth/token", body, merged)
  return AccessTokenSchema.parse(raw).access_token
}

/**
 * Disconnect the user's OAuth grant for `provider`. Best-effort upstream
 * revoke + local soft-delete. Returns true on success.
 */
export async function disconnect(
  provider: string,
  opts: DisconnectOptions = {},
): Promise<boolean> {
  if (typeof provider !== "string" || provider.length === 0) {
    throw new Error("provider must be a non-empty string")
  }
  const { userId, ...http } = opts
  let path = `/oauth/${encodeURIComponent(provider)}`
  if (userId !== undefined && userId !== "") {
    path += `?user_id=${encodeURIComponent(userId)}`
  }
  const raw = await request<{ disconnected?: boolean } | null>(
    "DELETE",
    path,
    null,
    http,
  )
  if (raw === null || raw === undefined) return true
  return Boolean(raw.disconnected ?? true)
}

/** Namespace object — `suite.oauth.*`. */
export const oauth = {
  authorizeUrl,
  connected,
  token,
  disconnect,
} as const

File: packages/sdk-py/af_stack/oauth.py
# SPDX-License-Identifier: Apache-2.0

"""``suite.oauth.*`` — OAuth-on-behalf-of-user operations.

Lets backend agents act on behalf of the user in third-party APIs
without re-implementing the OAuth dance per provider. The first shipped
providers are Google and GitHub. The runtime owns the dance + token
vault; this SDK is the client surface.

The four verbs:

============================  ====================================================
SDK call                      Endpoint
============================  ====================================================
:func:`authorize_url`         ``POST /api/v1/oauth/{provider}/authorize``
:func:`connected`             ``GET  /api/v1/oauth/connections``
:func:`token`                 ``POST /api/v1/oauth/token`` (internal-only header)
:func:`disconnect`            ``DELETE /api/v1/oauth/{provider}``
============================  ====================================================

:func:`token` is gated by the ``X-AF-Stack-Internal: 1`` header — browser
callers cannot reach it (the header is not in the CORS allow-list). The
runtime returns a freshly-refreshed access token; this SDK never sees a
refresh token.

The :class:`Connection` model carries metadata only — no token bytes.
The plaintext access token ONLY appears in :func:`token` responses.
"""

from __future__ import annotations

from urllib.parse import quote

from pydantic import BaseModel, ConfigDict, Field

from . import _http


class Connection(BaseModel):
    """One stored OAuth grant for the current user.

    Mirrors the Go ``oauth.Connection`` struct in
    ``services/runtime/internal/oauth/interface.go``.
    """

    model_config = ConfigDict(extra="allow")

    provider: str
    scopes: list[str] = Field(default_factory=list)
    connected_at: str
    expires_at: str | None = None


class ConnectionList(BaseModel):
    """Page of connections for the current user."""

    model_config = ConfigDict(extra="allow")

    connections: list[Connection] = Field(default_factory=list)


class AccessToken(BaseModel):
    """Plaintext access token + scopes.

    Only returned by :func:`token`. Refresh tokens NEVER cross the API
    surface — they live inside the secrets vault.
    """

    model_config = ConfigDict(extra="allow")

    provider: str
    user_id: str
    access_token: str
    scopes: list[str] = Field(default_factory=list)
    expires_at: str | None = None


async def authorize_url(
    provider: str,
    *,
    scopes: list[str] | None = None,
    return_to: str | None = None,
) -> str:
    """Return the provider's consent URL the user should visit.

    The runtime signs an HMAC state nonce that round-trips through the
    consent page; on callback it's verified to defeat CSRF.
    ``return_to`` is validated against an allowed-origin list — pass an
    origin on your customer-app (defaults match localhost dev origins).
    """
    if not provider:
        raise ValueError("provider is required")
    payload: dict[str, object] = {}
    if scopes is not None:
        payload["scopes"] = scopes
    if return_to is not None:
        payload["return_to"] = return_to
    body = await _http.request_json(
        "POST",
        f"/oauth/{quote(provider, safe='')}/authorize",
        json=payload,
    )
    if not isinstance(body, dict) or "authorization_url" not in body:
        raise _http.AFStackError(
            code="BAD_RESPONSE",
            message="authorize endpoint returned no authorization_url",
            status_code=200,
        )
    url = body["authorization_url"]
    if not isinstance(url, str):
        raise _http.AFStackError(
            code="BAD_RESPONSE",
            message="authorization_url is not a string",
            status_code=200,
        )
    return url


async def connected() -> ConnectionList:
    """List active OAuth grants for the current user."""
    body = await _http.request_json("GET", "/oauth/connections")
    return ConnectionList.model_validate(body or {"connections": []})


async def token(provider: str, user_id: str | None = None) -> str:
    """Return a fresh access token for ``provider`` (auto-refreshes).

    Internal-only — the runtime gates this endpoint with the
    ``X-AF-Stack-Internal: 1`` header. Workload modules and agents
    using the AF Stack SDK satisfy that header automatically; browser
    callers cannot.

    ``user_id`` overrides the resolved user. Pass it when calling under
    an API key (no user context) to fetch a token on behalf of a
    specific user the agent is serving.
    """
    if not provider:
        raise ValueError("provider is required")
    payload: dict[str, object] = {"provider": provider}
    if user_id is not None:
        payload["user_id"] = user_id
    body = await _http.request_json(
        "POST",
        "/oauth/token",
        json=payload,
        headers={"X-AF-Stack-Internal": "1"},
    )
    tok = AccessToken.model_validate(body or {})
    if not tok.access_token:
        raise _http.AFStackError(
            code="BAD_RESPONSE",
            message=f"token endpoint returned no access_token for {provider!r}",
            status_code=200,
        )
    return tok.access_token


async def disconnect(provider: str, user_id: str | None = None) -> bool:
    """Disconnect the user's grant for ``provider``.

    Best-effort upstream revoke + soft-delete locally. Returns ``True``
    on success.
    """
    if not provider:
        raise ValueError("provider is required")
    path = f"/oauth/{quote(provider, safe='')}"
    if user_id:
        path += f"?user_id={quote(user_id, safe='')}"
    body = await _http.request_json("DELETE", path)
    if not isinstance(body, dict):
        return True
    return bool(body.get("disconnected", True))


__all__ = [
    "AccessToken",
    "Connection",
    "ConnectionList",
    "authorize_url",
    "connected",
    "disconnect",
    "token",
]
